graphrag/index/flows/create_final_communities.py

```diff
@@ -4,82 +4,71 @@
 """All the steps to transform final communities."""
 
 from datetime import datetime, timezone
-from typing import cast
 from uuid import uuid4
 
 import pandas as pd
-from datashaper import (
-    VerbCallbacks,
-)
-
-from graphrag.index.operations.unpack_graph import unpack_graph
 
 
 def create_final_communities(
-    entity_graph: pd.DataFrame,
-    callbacks: VerbCallbacks,
+    base_entity_nodes: pd.DataFrame,
+    base_relationship_edges: pd.DataFrame,
+    base_communities: pd.DataFrame,
 ) -> pd.DataFrame:
     """All the steps to transform final communities."""
-    graph_nodes = unpack_graph(entity_graph, callbacks, "clustered_graph", "nodes")
-    graph_edges = unpack_graph(entity_graph, callbacks, "clustered_graph", "edges")
-
-    # Merge graph_nodes with graph_edges for both source and target matches
-    source_clusters = graph_nodes.merge(
-        graph_edges, left_on="label", right_on="source", how="inner"
+    # aggregate entity ids for each community
+    entity_ids = base_communities.merge(base_entity_nodes, on="title", how="inner")
+    entity_ids = (
+        entity_ids.groupby("community").agg(entity_ids=("id", list)).reset_index()
     )
 
-    target_clusters = graph_nodes.merge(
-        graph_edges, left_on="label", right_on="target", how="inner"
+    # aggregate relationships ids for each community
+    # these are limited to only those where the source and target are in the same community
+    max_level = base_communities["level"].max()
+    all_grouped = pd.DataFrame(
+        columns=["community", "level", "relationship_ids", "text_unit_ids"]  # type: ignore
     )
-
-    # Concatenate the source and target clusters
-    clusters = pd.concat([source_clusters, target_clusters], ignore_index=True)
-
-    # Keep only rows where level_x == level_y
-    combined_clusters = clusters[
-        clusters["level_x"] == clusters["level_y"]
-    ].reset_index(drop=True)
-
-    cluster_relationships = (
-        combined_clusters.groupby(["cluster", "level_x"], sort=False)
-        .agg(
-            relationship_ids=("id_y", "unique"),
-            text_unit_ids=("source_id_x", "unique"),
-            entity_ids=("id_x", "unique"),
+    for level in range(max_level + 1):
```

Comment:
If we start from 0 here we can effectively filter out all the -1 communities.
I'm currently filtering them out on node prep for community reports as well.

Reply:
This is starting from level 0. You may be confusing the community ID with the level? IDs can be -1 for nodes when they have no community home at a given level.